lib/policy_machine_storage_adapters/active_record.rb

```diff
@@ -796,6 +796,49 @@ def accessible_objects(user_or_attribute, operation, options = {})
       end
     end
 
+    # Version of accessible_objects which only returns objects that are
+    # ancestors of a specified root object or the object itself.
+    def accessible_ancestor_objects(user_or_attribute, operation, root_object, options = {})
+      # The final set of accessible objects must be contained in the following set
+      ancestor_objects = root_object.ancestors(type: class_for_type('object')) + [root_object.stored_pe]
+
+      # Fetch all of the PEAs using the given UA or its descendants
+      user_attribute_ids = user_or_attribute.descendants.pluck(:id) | [user_or_attribute.id]
```

Comment:
There's a lot of code duplication between this and `accessible_objects` which could be DRYed up.